Improved error handling.

goyox86 opened the pull request:
So basically this is the first take on error handling. I decided to give it a try to [failure](https://crates.io/crates/failure) and I think is awesome.

Anyways I'm still not happy with the error enum. Might be too detailed. There is still a case in which I neeed to see *why* something failed. Basically the error I get from the library is a trait object (`failure::Error`) and I have to downcast to act on the error type, and that's the part I'm not sure yet, I need to continue exploring `failure` to see if I can improve the usability whenever you want to act based on the types of errors. But looks promising! at least from the library that generates the errors, it is much much nicer than the `Error + Debug + Display + From` combo!

Just take a look @jackpot51 and @MggMuggins lets iterate this PR! \o/  

Ah! I've migrated `userutils` locally to this and seems to be working fine look at the diff [diff](https://github.com/redox-os/userutils/compare/master...goyox86:goyox86/better-errors?expand=1)

goyox86 committed:
Improved error handling.

TODO: Describe the details of the changes.

jackpot51 commented:
It is looking much better with these custom error types.

MggMuggins commented:
We could commit this to a new branch here for a simpler iterative dev cycle.

#### As per your changes to redox_users:
Anyway, I think this looks reasonable (and cleaner). I definitely like it better than what it was, given that it won't ever just crash things...
I notice that you created a bunch of functions to easily create different kinds of errors. I see why you did it, and it makes sense, but I'm not sure I like it ( <- Not a good reason not to do it...). Other than that, this makes me happy. :P

#### As per the userutils port:
I noticed at least one issue with your edits to userutils. Lines 75-77 were removed. See the `--force` flag in the man page to see if you can figure out why that shouldn't have happened. ;D

Seems to me that the changes you made to `login` illustrate the only downside to this method: some added verbosity for certain kinds are error handling, specifically lots of nested matches... I'm not sure that I can read that code. Maybe worth trying to find a better solution.

Something else occurred to me while reading through this, see https://github.com/redox-os/arg-parser/issues/5.

goyox86 committed:
Simplified error types and bit of refactoring.

- Made a type alias Result<T>.
- Removed a few variant of the NotFound and AlreadyExists errors.
- Updated to failure v1.1.1.
- Removed some helper methods as building errors is simpler now.

MggMuggins commented:
Kudos on that! I've opened a PR to libextra (https://github.com/redox-os/libextra/pull/10) for `fail_because`, which I think will make error handling simpler in all the utils crates.

MggMuggins commented:
If we can figure out what packages we're breaking with these changes, could we potentially get these merged? I'd like to overhaul API and this would be really helpful.

goyox86 commented:
@jackpot51 Whenever you can take a look at this 🙏 And let us know if you are OK with merging :)

jackpot51 commented:
It looks good now - I think it is ready to merge.
